muxers/mplex/src/io.rs

```diff
@@ -346,37 +352,96 @@ where
         self.guard_open()?;
 
         // Try to read from the buffer first.
-        if let Some((pos, data)) = self.buffer.iter()
-            .enumerate()
-            .find_map(|(pos, frame)| match frame {
-                Frame::Data { stream_id, data }
-                    if stream_id.into_local() == id => Some((pos, data.clone())),
-                _ => None
-            })
-        {
-            if self.buffer.len() == self.config.max_buffer_len {
-                // The buffer is full and no longer will be, so notify all pending readers.
-                ArcWake::wake_by_ref(&self.notifier_read);
+        if let Some(state) = self.substreams.get_mut(&id) {
+            let buf = state.recv_buf();
+            if !buf.is_empty() {
+                if self.blocked_stream == Some(id) {
+                    // Unblock reading new frames.
+                    self.blocked_stream = None;
+                    ArcWake::wake_by_ref(&self.notifier_read);
+                }
+                let data = buf.remove(0);
+                return Poll::Ready(Ok(Some(data)))
             }
-            self.buffer.remove(pos);
-            return Poll::Ready(Ok(Some(data)));
+            // If the stream buffer "spilled" onto the heap, free that memory.
+            buf.shrink_to_fit();
         }
 
         loop {
             // Check if the targeted substream (if any) reached EOF.
             if !self.can_read(&id) {
+                // Note: Contrary to what is recommended by the spec, we must
+                // return "EOF" also when the stream has been reset by the
+                // remote, as the `StreamMuxer::read_substream` contract only
+                // permits errors on "terminal" conditions, e.g. if the connection
+                // has been closed or on protocol misbehaviour.
                 return Poll::Ready(Ok(None))
             }
 
+            // Check if there is a blocked stream.
+            if let Some(blocked_id) = &self.blocked_stream {
+                // We have a blocked stream and cannot continue reading
+                // new frames for any stream, until frames are taken from
+                // the blocked stream's buffer. Try to wake pending readers
+                // of the blocked stream.
+                if !self.notifier_read.wake_read_stream(*blocked_id) {
+                    // No task dedicated to the blocked stream woken, so schedule
+                    // this task again to have a chance at progress.
+                    cx.waker().clone().wake();
+                } else if blocked_id != &id {
```

Comment:
```suggestion
                } else {
                    if blocked_id == &id {
                        debug_assert!(false, "...");
                    }     
                }
```

What do you think of ensuring that a read on a blocking stream never makes it this far?

Reply:
Why do you think it can never happen that the current task is woken at line `387`?

Reply:
I do think it can happen that the current task is woken up in `387` and thus control flow reaches into the `else` block in `391`, but only when the current task is not trying to read from a stream that is currently blocking the connection.

My suggestion is confusing. I hope I am not missing something. Let me retry:

Say the read buffer of stream _S1_ is full and thus is blocking the entire connection. Say that the current task is trying to read from _S1_. In this case the current task would return `Poll::Ready(Ok(Some(data)))` in `364` already before the `loop`.

With the above in mind, one can not reach the loop when trying to read from the stream (in the case above _S1_) that is currently blocking the connection. To enforce this one could insert a `debug_assert` below `382` like such:

```diff
             if let Some(blocked_id) = &self.blocking_stream {
+               if blocked_id == id { debug_assert!(false, "..."); }
+
                // We have a blocked stream and cannot continue reading
                // new frames for any stream, until frames are taken from
                // the blocked stream's buffer. Try to wake a pending reader
                // of the blocked stream.
```

I don't think this is particularly important. Thus feel free to ignore this suggestion. What I would like to ensure though is that I am not misunderstanding something.

Reply:
I see. Your comment made me aware though that, contrary to my own expectations, the check for the blocking stream is currently only in `poll_read_stream`. That means `poll_next_stream` is never blocked and can grow the substream buffers indefinitely, only limited by hitting the `max_substreams` limit. I corrected that [here](https://github.com/libp2p/rust-libp2p/pull/1784/commits/fde6f008d489abf259e2cc1978c7212167b4c37e) together with more commentary. I would appreciate if you could give that another critical look to see if it all makes sense now. I'll see if I can come up with some property tests w.r.t. the `MaxBufferBehaviour`.